haystack/document_store/sql.py

```diff
@@ -117,19 +117,42 @@ def get_all_documents(
         # Generally ORM objects kept in memory cause performance issue
         # Hence using directly column name improve memory and performance.
         # Refer https://stackoverflow.com/questions/23185319/why-is-loading-sqlalchemy-objects-via-the-orm-5-8x-slower-than-rows-via-a-raw-my
-        query = self.session.query(
+        documents_query = self.session.query(
             DocumentORM.id,
             DocumentORM.text,
-            DocumentORM.meta
+            DocumentORM.vector_id
         ).filter_by(index=index)
 
         if filters:
-            query = query.join(MetaORM)
+            documents_query = documents_query.join(MetaORM)
             for key, values in filters.items():
-                query = query.filter(MetaORM.name == key, MetaORM.value.in_(values))
+                documents_query = documents_query.filter(
+                    MetaORM.name == key,
+                    MetaORM.value.in_(values),
+                    DocumentORM.id == MetaORM.document_id
+                )
+
+        documents_map = {}
+        for row in documents_query.all():
+            documents_map[row.id] = Document(
+                id=row.id,
+                text=row.text,
+                meta=None if row.vector_id is None else {"vector_id": row.vector_id}
+            )
 
-        documents = [self._convert_sql_row_to_document(row) for row in query.all()]
-        return documents
+        if len(documents_map) > 0:
+            meta_query = self.session.query(
+                MetaORM.document_id,
+                MetaORM.name,
+                MetaORM.value
+            ).filter(MetaORM.document_id.in_(documents_map.keys()))
+
+            for row in meta_query.all():
```

Comment:
The first query (line 120) executes quickly as it should if just querying by index (< 1 sec). Memory usage seems to be normal / expected.

The second call to get metadata is much slower and errors eventually:
```
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) too many SQL variables
[SQL: SELECT meta.document_id AS meta_document_id, meta.name AS meta_name, meta.value AS meta_value
FROM meta
WHERE meta.document_id IN (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,.....?)
```

It seems odd that 2 separate queries would be required to get the same fields in a document.

Reply:
> The second call to get metadata is much slower and errors eventually:

To fix it I have added index on document_id and limiting the number of host variable parameters passed to sql queries. (999 is for sqlite < 3.32 and 32K for >= 3.32)

> It seems odd that 2 separate queries would be required to get the same fields in a document.

It is done to prevent duplication of very long `text` field in memory. Each docs can have multiple metas and for each meta better not to keep duplicate text in the memory. Hence I have split it in two queries.

Reply:
Can you please test with latest change where I have fix the issue you have reported.
It seems you have very good amount of data to benchmark it.

Reply:
@vinchg by chance if you get time could you please test latest changes.